metricq_tools/energy.py

```diff
@@ -0,0 +1,85 @@
+import asyncio
+import time
+
+import click
+import metricq
+
+from .logging import logger
+from .utils import TemplateStringParam, client_version, metricq_command, run_cmd
+
+
+async def energy(
+    subscriber: metricq.Subscriber, metric: str, command: list[str]
+) -> None:
+    async with subscriber:
+        # No need to keep the subscriber alive, this is just to setup the queue
+        pass
+    time_before = time.time()
+    await run_cmd(command)
+    duration = time.time() - time_before
+    assert duration > 0
+    value_sum = 0.0
+    value_count = 0
+    async with subscriber.drain() as data:
+        async for data_metric, timestamp, value in data:
+            assert data_metric == metric
+            value_sum += value
+            value_count += 1
+
+    click.echo(
+        f"[metricq-energy] duration={duration:.3f} s number of values={value_count}"
+    )
+
+    if value_count == 0:
+        logger.error(f"No data received for {metric}, cannot compute energy.")
+        return
+    if value_count < 10:
+        logger.warning(
+            f"Received only {value_count} data points for {metric}, "
+            f"energy computation may be inaccurate."
+        )
+    value_mean = value_sum / value_count
+    energy = value_mean * duration
+
+    click.echo(f"[metricq-energy] mean of {metric}={value_mean:.1f} W")
+    click.echo(f"[metricq-energy] integral of {metric}={energy:.1f} J")
+
+
+@metricq_command(default_token="sink-$USER-tool-energy")
+@click.option(
+    "-m", "--metric", type=TemplateStringParam(), required=True, multiple=False
+)
+@click.option(
+    "--expires",
+    type=int,
+    default=3600,
+    help="Queue expiration time in seconds. Set this value to the maximum time the "
+    "command is expected to run.",
+)
+@click.argument("command", required=True, nargs=-1)
+def main(
+    server: str,
+    token: str,
+    metric: str,
+    expires: int,
+    command: list[str],
+):
+    """
+    Get a single energy value for a metric during the execution of the given command.
+    This value is just the integral of the metric over the time the command was running.
+
+    If the metric is in watts (W), the result is in joules (J). The unit is not checked.
+
+    The computation uses the arithmetic mean of the metric values and the execution
```

Comment:
I only understood this sentence once I read the source code. Better something like:

> The integral is calculated as the product of the arithmetic mean of all values times the runtime of the given command.